### ISSUE TITLE
Can't use array values as indices in other arrays?

### ISSUE THREAD

**Issue by maconard** [grencez/protocon#14]
I'm attempting to use the value of an array item to index another array, but it seems this functionality doesn't exist?

Relevant code:
```
constant N := 3;
constant X := 20;
constant Y := 20;
constant DIM := X*Y;

variable c[DIM] < N+1;
variable x[N] < X;
variable y[N] < Y;

predicate M := (forall i <- Nat % DIM : c[i] == N || (c[i] < N && x[c[i]]+X*y[c[i]] == i)); //Line 56
```

Error:
```
./protocon/prot-xfile.cc(1558) eval_int: 
./protocon/prot-xfile.cc(14) update_allgood: Something terribly wrong at/after line:56
./protocon/prot-xfile.cc(1644) lookup_vbl: eval array index
./protocon/prot-xfile.cc(1539) eval: 
./protocon/prot-xfile.cc(1486) eval: 
./protocon/prot-xfile.cc(1279) eval: 
./protocon/prot-xfile.cc(1215) eval: 
protocon: /home/michael/Documents/Software/protocon/bld/bld/protocon/prot-xfile.cc:1210: bool protocon::ProtoconFile::eval(protocon::P::Fmla&, Sesp): Assertion `0' failed.
Aborted (core dumped)
```

So my question is, is there a different syntax for what I'm attempting or is what I'm attempting simply not allowed?

Thanks,

**Comment by grencez** [comment]
You're right, variables can't be used to index arrays. There should probably be an error message and documentation for this. Basically it's because the tool only supports static topologies (the readable and writable variables of a process cannot change at runtime).

The workaround for predicates is to introduce another quantifier. Your predicate would be `(forall i <- Nat % DIM: (forall j <- Nat % N: j==c[i] ==> x[j]+X*y[j]==i))`, where `j` is effectively bound to `c[i]`.

**Comment by maconard** [comment]
Interesting. Thank you for that workaround snippet.

> Basically it's because the tool only supports static topologies (the readable and writable variables of a process cannot change at runtime).

Well, I am trying to do a dynamic topology... We'll see if I can find a way to trick Protocon into working for me. Thinking about allowing read/write for the values of all nodes and then restricting that at runtime via some other quality, to get around not being able to change read/write at runtime.

Thanks!

**Comment by grencez** [comment]
So an analogy to the `predicate` way of indexing arrays with other variables would look like this:

```
process P[id < N*DIM]
{
  let i := id % N;
  let z := id / N;
  write: x[i], y[i], c[z];  // implies read also

  predicate ValidZ := (z == x[i] + X*y[i]);
  // Only let this process act when Z is the index you want.
  forbid: ( !ValidZ -> );
}
```

It doesn't actually work because processes are not allowed to act differently based on their indices, making the `ValidZ` predicate invalid. I could introduce a syntax to support this. Something like a `read: Constant[z];` permission to read a constant integer "variable" that has the value `z`. Weird but it might work.

But I fear you're headed towards state explosion either way. So before implementing that questionable feature, could you make sure it actually gives you something useful? You can achieve the exact same effect now in a hacky way by making a different process type for each `z` value like so:

```
process P0[i < N]
{
  let z := 0;
  write: x[i], y[i], c[z];
  predicate ValidZ := (0 == x[i] + X*y[i]);
  forbid: ( !ValidZ -> );
}

process P1[i < N]
{
  let z := 1;
  write: x[i], y[i], c[z];
  predicate ValidZ := (1 == x[i] + X*y[i]);
  forbid: ( !ValidZ -> );
}

process P2[i < N]
{
  let z := 2;
  write: x[i], y[i], c[z];
  predicate ValidZ := (2 == x[i] + X*y[i]);
  forbid: ( !ValidZ -> );
}

// etc.
```

It shouldn't be *too* hard to make a script generate the duplicate processes.

As for the state explosion concern, hopefully you can find a meaningful model with small variables. For example, binary or ternary can be a pretty decent abstraction when your real code will just be swapping values around based on "less than", "greater than", and "equal to" comparisons. It doesn't look like you're doing that here... but just FYI, even a couple writable Nat%20 variables will blow up the solver. :)
